### PR TITLE
redirect_uri_integrity_check_bypass

### PR DESCRIPTION
**Pull request by taitzariel**

### CODE REVIEW

File: server/src/main/java/org/cloudfoundry/identity/uaa/oauth/AntPathRedirectResolver.java [hunk 1]
```diff
@@ -67,6 +69,44 @@ protected boolean redirectMatches(String requestedRedirect, String clientRedirec
...
+     *     checks a match for:
+     *     <ol>
+     *         <li>example.com/foo/%252e./bar</li>
+     *         <li>example.com/foo/%2e./bar</li>
+     *         <li>example.com/foo/../bar</li>
+     *         <li>example.com/bar</li>
+     *     </ol>
+     * </p>
+     */
+    private boolean matches(Predicate<String> matcher, String requestedRedirect) {
```
Comment (metacubed) [inline server/src/main/java/org/cloudfoundry/identity/uaa/oauth/AntPathRedirectResolver.java]: Is there any way to separate this into two functions - one to recursively decode, and another to do the matching?
If not, please rename this function to better reflect the functionality, something like `decodeRecursivelyAndMatch` (just an example, please don't use that one 😄).

File: server/build.gradle [hunk 2]
```diff
@@ -3,7 +3,9 @@ description = 'CloudFoundry Identity Server JAR'
 dependencies {
 
   compile project(':cloudfoundry-identity-model')
-
+  compile group: 'org.apache.tomcat.embed', name: 'tomcat-embed-core', version:parent.tomcatVersion
+  compileOnly group: 'org.projectlombok', name: 'lombok', version: parent.lombokVersion
```
Comment (metacubed) [inline server/build.gradle]: Please check whether you still need `compileOnly` if you already have `annotationProcessor` on the same dependency.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/ImplicitGrantIT.java [hunk 3]
```diff
@@ -62,34 +59,13 @@
...
-
     @Autowired
     RestOperations restOperations;
 
-    @Autowired
-    TestClient testClient;
-
-    @Before
-    @After
-    public void logout_and_clear_cookies() {
```
Comment (metacubed) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/ImplicitGrantIT.java]: What is the purpose of this change?

File: server/src/main/java/org/cloudfoundry/identity/uaa/oauth/AntPathRedirectResolver.java [hunk 4]
```diff
@@ -88,6 +128,10 @@ public String resolveRedirect(String requestedRedirect, ClientDetails client) th
         return super.resolveRedirect(requestedRedirect, client);
     }
 
+    private static boolean isWildcard(String configuredRedirectPattern) {
```
[the remark this answers is not in the document]
Comment (taitzariel) [reply server/src/main/java/org/cloudfoundry/identity/uaa/oauth/AntPathRedirectResolver.java]: The function had needlessly been a member of the inner class.
I moved it to simplify code readability.
Would you like me to move it back, nonetheless?

File: server/src/main/java/org/cloudfoundry/identity/uaa/oauth/AntPathRedirectResolver.java [hunk 5]
```diff
@@ -67,6 +69,44 @@ protected boolean redirectMatches(String requestedRedirect, String clientRedirec
...
+     *     <ol>
+     *         <li>example.com/foo/%252e./bar</li>
+     *         <li>example.com/foo/%2e./bar</li>
+     *         <li>example.com/foo/../bar</li>
+     *         <li>example.com/bar</li>
+     *     </ol>
+     * </p>
+     */
+    private boolean matches(Predicate<String> matcher, String requestedRedirect) {
+        for (int i = 1; i <= 5; i++) {
```
[the remark this answers is not in the document]
Comment (taitzariel) [reply server/src/main/java/org/cloudfoundry/identity/uaa/oauth/AntPathRedirectResolver.java]: Yes, it's required. The explanation is in the log.debug() message.
I've now given it a meaningful name.

File: server/src/main/java/org/cloudfoundry/identity/uaa/oauth/AntPathRedirectResolver.java [hunk 6]
```diff
@@ -67,6 +69,44 @@ protected boolean redirectMatches(String requestedRedirect, String clientRedirec
...
+            if (!matcher.test(requestedRedirect)) {
+                return false;
+            }
+            String decoded = urlDecode(requestedRedirect);
+            if (decoded.equals(requestedRedirect)) {
+                return matcher.test(StringUtils.cleanPath(decoded));
+            }
+            requestedRedirect = decoded;
+        }
+        logger.debug("aborted url decoding loop to mitigate DOS attack that sends a repeatedly url-encoded path");
```
[the remark this answers is not in the document]
Comment (taitzariel) [reply server/src/main/java/org/cloudfoundry/identity/uaa/oauth/AntPathRedirectResolver.java]: OK. I've capitalized it. Out of a sample of log messages that I looked at, the majority do not end with a period.

File: server/src/main/java/org/cloudfoundry/identity/uaa/oauth/AntPathRedirectResolver.java (hunk 1 above)
Comment (taitzariel) [reply server/src/main/java/org/cloudfoundry/identity/uaa/oauth/AntPathRedirectResolver.java]: I'm not sure if it can be split in a way that would enhance readability.
I've renamed the matches() function.

File: server/build.gradle (hunk 2 above)
Comment (taitzariel) [reply server/build.gradle]: I've checked. It will not compile without the "compileOnly" line.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/ImplicitGrantIT.java (hunk 3 above)
Comment (taitzariel) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/ImplicitGrantIT.java]: Before I wrote the unit tests, I was considering writing an integration test. This is one of the classes that I considered writing the integration test in. I didn't understand the purpose of various members of this class, like these, and then found that in fact they were unused.
Hence I removed them to save confusion for the next developer who will read this class.

**Comment by taitzariel** [comment]
>     * You have commits under two different users - is that intentional? One of those users is not known to GitHub.
I had committed from different computers. I then tried to understand how to make the username  consistent with the "git username test" commits.

**Comment by taitzariel** [comment]
>     * Simplify commit history by squashing intermediate commits which will clutter up master. For example,
OK. Done.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/ImplicitGrantIT.java (hunk 3 above)
Comment (metacubed) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/ImplicitGrantIT.java]: Please **revert this change** even if it is useful. GE does not want to have any differences from the Pivotal codebase, unless they are related to some GE-specific features.

If you want, you can make a separate PR for this file directly against Pivotal UAA. If they approve it, then we will get the changes during the next version upgrade.

Comment (metacubed) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/ImplicitGrantIT.java]: Please **revert this change** even if it is useful. GE does not want to have any differences from the Pivotal codebase, unless they are related to some GE-specific features.

If you want, you can make a separate PR for this file directly against Pivotal UAA. If they approve it, then we will get the changes during the next version upgrade.

File: server/src/main/java/org/cloudfoundry/identity/uaa/oauth/AntPathRedirectResolver.java [hunk 7]
```diff
@@ -67,6 +69,45 @@ protected boolean redirectMatches(String requestedRedirect, String clientRedirec
...
+     *     <ol>
+     *         <li>example.com/foo/%252e./bar</li>
+     *         <li>example.com/foo/%2e./bar</li>
+     *         <li>example.com/foo/../bar</li>
+     *         <li>example.com/bar</li>
+     *     </ol>
+     * </p>
+     */
+    private boolean matchesAfterNormalization(Predicate<String> matcher, String requestedRedirect) {
+        int maxDecodeAttempts = 5;
```
Comment (metacubed) [inline server/src/main/java/org/cloudfoundry/identity/uaa/oauth/AntPathRedirectResolver.java]: Make this a Java constant please - at the start of the class:
```java
private static final int MAX_DECODE_ATTEMPTS = 5
```

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/ImplicitGrantIT.java (hunk 3 above)
Comment (taitzariel) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/ImplicitGrantIT.java]: OK

File: server/src/main/java/org/cloudfoundry/identity/uaa/oauth/AntPathRedirectResolver.java (hunk 7 above)
Comment (taitzariel) [reply server/src/main/java/org/cloudfoundry/identity/uaa/oauth/AntPathRedirectResolver.java]: It is used in this method only. Why make it a field?
